migration/migrator/migrations/course/20240205154116_add_poll_timer.py

```diff
@@ -0,0 +1,49 @@
+"""Migration for a given Submitty course database."""
+
+
+def up(config, database, semester, course):
+    """
+    Run up migration.
+
+    :param config: Object holding configuration details about Submitty
+    :type config: migrator.config.Config
+    :param database: Object for interacting with given database for environment
+    :type database: migrator.db.Database
+    :param semester: Semester of the course being migrated
+    :type semester: str
+    :param course: Code of course being migrated
+    :type course: str
+    """
+    database.execute("ALTER TABLE polls ADD COLUMN IF NOT EXISTS duration INTEGER DEFAULT 0")
+    database.execute("ALTER TABLE polls ADD COLUMN IF NOT EXISTS end_time timestamptz DEFAULT NULL")
+    database.execute("ALTER TABLE polls ALTER COLUMN status DROP NOT NULL")
+    database.execute("ALTER TABLE polls ADD COLUMN IF NOT EXISTS is_visible BOOLEAN NOT NULL DEFAULT FALSE")
+    database.execute("""
+        UPDATE polls SET 
+            end_time = CASE
+                WHEN status = 'open' THEN NULL
+                ELSE now() 
+            END,
+            is_visible = CASE
+                WHEN status IN ('open', 'ended') THEN TRUE
+                ELSE FALSE
+            END
+    """)
+
+
+def down(config, database, semester, course):
+    """
+    Run down migration (rollback).
+
+    :param config: Object holding configuration details about Submitty
+    :type config: migrator.config.Config
+    :param database: Object for interacting with given database for environment
+    :type database: migrator.db.Database
+    :param semester: Semester of the course being migrated
+    :type semester: str
+    :param course: Code of course being migrated
+    :type course: str
+    """
+    pass
```

Comment:
Is there a safe down migration which can be applied here to ensure a reasonable status is set in case this gets rolled back for whatever reason?

Reply:
status columns wouldn't ever change. So rolling it back should be fine.

Reply:
idk if there is a need to have a down migration

Reply:
Never mind, I made the down migration reset the status constraint back to the original.

Reply:
I believe the down migration leaves the system in a broken state if new polls are created with these changes, and then the changes are rolled back.  The new polls will not have a valid status.  Now that you've added the constraint back, I believe the down migration itself will fail due to the null values in new polls.

This migration needs to be 100% reversible and repeatable in both directions.

Reply:
I think I understand